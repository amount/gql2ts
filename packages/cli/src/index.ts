#!/usr/bin/env node
'use strict';
import * as program from 'commander';
import * as fs from 'fs';
import {
  readFile,
  writeToFile,
  PossibleSchemaInput,
  safeJSONParse,
} from '@gql2ts/util';
import { ISchemaToInterfaceOptions, generateNamespace } from '@gql2ts/from-schema';
import fromQuery from '@gql2ts/from-query';
import { IFromQueryReturnValue } from '@gql2ts/types';

// tslint:disable-next-line no-require-imports no-var-requires
const { version } = require('../package.json');

program
  .version(version)
  .usage('[options] <schema.json | schema.gql> <query.gql>')
  .option('-o --output-file [outputFile]', 'name for output file, will use stdout if not specified')
  .option('-n --namespace [namespace]', 'name for the namespace, defaults to "GQL"', 'GQL')
  .option('-i --ignored-types <ignoredTypes>', 'names of types to ignore (comma delimited)', v => v.split(','), [])
  .option('-l --legacy', 'Use TypeScript 1.x annotation', false)
  .option('-e --external-options [externalOptions]', 'ES Module with method overwrites')
  .option('--ignore-type-name-declaration', 'Whether to exclude __typename', false)
  .parse(process.argv);

interface ICLIOptions extends Partial<ISchemaToInterfaceOptions> {
  queryString: string;
}

const run: (schema: PossibleSchemaInput, options: Partial<ICLIOptions>) => void = (schema, options) => {
  let defaultOverrides: object = {};
  if (program.externalOptions) {
    // tslint:disable-next-line no-require-imports no-var-requires
    const externalFile: any = require(program.externalOptions);
    defaultOverrides = externalFile.default || externalFile;
  }

  if (program.args[1]) {
    const queryFile: string = program.args[1];
    const query: string = fs.readFileSync(queryFile).toString();
    const info: IFromQueryReturnValue[] = fromQuery(schema, query, {}, defaultOverrides);
    const toWrite: string = info.map(inf => inf.result).join('\n\n');
    if (options.outputFile) {
      writeToFile(options.outputFile, toWrite);
    } else {
      console.log(toWrite);
    }
    return;
  }

  let namespace: string = generateNamespace(options.namespace!, schema, options, defaultOverrides);

  if (options.outputFile) {
    writeToFile(options.outputFile, namespace);
  } else {
    console.log(namespace);
  }
};

const fileName: string | undefined = program.args[0];

if (fileName) {
  const schema: string | object = readFile(fileName);
  run(schema as PossibleSchemaInput, program as any);
} else if (!process.stdin.isTTY) {
  let input: string = '';
  process.stdin.resume();
  process.stdin.setEncoding('utf8');
  process.stdin.on('data', (data) => {
    input += data;
  });
  process.stdin.on('end', () => run(safeJSONParse(input) as PossibleSchemaInput, program as any));
} else {
  console.error('No input specified. Please use stdin or a file name.');
  program.outputHelp();
}
