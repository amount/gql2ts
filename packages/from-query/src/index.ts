import {
  parse,
  GraphQLSchema,
  DocumentNode,
  GraphQLType,
  VariableDefinitionNode,
  TypeNode,
  GraphQLEnumType,
  GraphQLInputObjectType,
  GraphQLInputField,
} from 'graphql';
import {
  schemaFromInputs,
  isList,
  isNonNullable,
  isEnum,
  filterAndJoinArray,
} from '@gql2ts/util';
import {
  IFromQueryOptions,
  HandleNamedTypes,
  HandleInputTypes,
  ConvertToTypeSignature,
  FromQuerySignature,
} from '@gql2ts/types';
import {
  DEFAULT_TYPE_MAP,
  DEFAULT_OPTIONS,
} from '@gql2ts/language-typescript';
import { flattenFragments } from './flattenQuery';
import convertToIr, { IOperation } from './ir';
import generateTypes from './generate';

const doIt: FromQuerySignature = (schema, query, typeMap = {}, providedOptions = {}) => {
  const options: IFromQueryOptions = {
    ...DEFAULT_OPTIONS,
    ...providedOptions,
    typeMap: {
      ...DEFAULT_TYPE_MAP,
      ...DEFAULT_OPTIONS.typeMap,
      ...(providedOptions.typeMap ? providedOptions.typeMap : {}),
      ...typeMap
    }
  };

  const {
    wrapList,
    printType,
    formatInput,
    generateQueryName,
    interfaceBuilder,
    generateInterfaceDeclaration,
    exportFunction,
    postProcessor,
    generateInputName,
    typeMap: TypeMap,
  }: IFromQueryOptions = options;

  const parsedSchema: GraphQLSchema = schemaFromInputs(schema);
  const parsedSelection: DocumentNode = parse(query);

  const handleInputObject: (type: GraphQLInputObjectType, isNonNull: boolean) => string = (type, isNonNull) => {
    const variables: GraphQLInputField[] = Object.keys(type.getFields()).map(k => type.getFields()[k]);
    const variableDeclarations: string[] = variables.map(v => {
      // TODO fix this
      const convertedType: string = convertToType(v.type);

      return formatInput(v.name, convertedType.endsWith('null'), convertedType);
    });
    const builder: string = generateInterfaceDeclaration(variableDeclarations);
    return printType(builder, isNonNull);
  };

  const typeUnion: (types: string[]) => string = types => types.join(' | ');

  const handleEnum: (type: GraphQLEnumType, isNonNull: boolean) => string = (type, isNonNull) => (
    printType(
      typeUnion(type.getValues().map(({ value }) => `'${value}'`)),
      isNonNull
    )
  );

  const handleNamedTypeInput: (type: TypeNode, isNonNull: boolean) => string | undefined = (type, isNonNull) => {
    if (type.kind === 'NamedType' && type.name.kind === 'Name' && type.name.value) {
      const newType: GraphQLType | undefined | null = parsedSchema.getType(type.name.value);
      if (newType instanceof GraphQLEnumType) {
        return handleEnum(newType, isNonNull);
      } else if (newType instanceof GraphQLInputObjectType) {
        return handleInputObject(newType, isNonNull);
      }
    }
  };

  const handleRegularType: HandleNamedTypes = (type, isNonNull, replacement) => {
    const typeValue: string = (typeof type.name === 'string') ? type.toString() : type.name.value;
    const showValue: string = replacement || typeValue;
    const show: string = TypeMap[showValue] || (replacement ? showValue : TypeMap.__DEFAULT);
    return printType(show, isNonNull);
  };

  const convertVariable: HandleInputTypes = (type, isNonNull = false, replacement = null) => {
    if (type.kind === 'ListType') {
      return printType(wrapList(convertVariable(type.type, false, replacement)), isNonNull!);
    } else if (type.kind === 'NonNullType') {
      return convertVariable(type.type, true, replacement);
    } else {
      return handleNamedTypeInput(type, isNonNull!) || handleRegularType(type, isNonNull!, replacement!);
    }
  };

  const convertToType: ConvertToTypeSignature = (type, isNonNull = false, replacement = null) => {
    if (isList(type)) {
      return printType(wrapList(convertToType(type.ofType, false, replacement)), isNonNull!);
    } else if (isNonNullable(type)) {
      return convertToType(type.ofType, true, replacement);
    } else if (isEnum(type)) {
      return handleEnum(type, isNonNull!);
    } else if (type instanceof GraphQLInputObjectType) {
      return handleInputObject(type, isNonNull);
    } else {
      return handleRegularType(type, isNonNull!, replacement!);
    }
  };

  const getVariables: (variables: ReadonlyArray<VariableDefinitionNode>) => string[] = variables => (
    variables.map(v => {
      const optional: boolean = v.type.kind !== 'NonNullType';
      return formatInput(v.variable.name.value, optional, convertVariable(v.type));
    })
  );

  const variablesToInterface: (operationName: string, variables: ReadonlyArray<VariableDefinitionNode> | undefined) => string = (
    opName,
    variables
  ) => {
    if (!variables || !variables.length) { return ''; }

    const variableTypeDefs: string[] = getVariables(variables);
    return exportFunction(interfaceBuilder(generateInputName(opName), generateInterfaceDeclaration(variableTypeDefs)));
  };

  const flattenedQuery: DocumentNode = flattenFragments(parsedSelection, parsedSchema);
  const internalRepresentation: IOperation = convertToIr(parsedSchema, flattenedQuery);

  const variableInterfaces: string[] = parsedSelection.definitions.map(def => {
    if (def.kind !== 'OperationDefinition') { return ''; }

    const ifaceName: string = generateQueryName(def);
    return variablesToInterface(ifaceName, def.variableDefinitions);
  });

  return postProcessor(
    filterAndJoinArray([
      generateTypes(options)(internalRepresentation),
      '\n',
      ...variableInterfaces
    ])
  );
};

export default doIt;
