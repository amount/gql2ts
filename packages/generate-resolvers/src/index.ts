import {
  GraphQLSchema,
  GraphQLObjectType,
  GraphQLFieldMap,
  GraphQLField,
} from 'graphql';

import {
  schemaFromInputs,
  PossibleSchemaInput,
} from '@gql2ts/util';
// import { IFromQueryOptions } from '@gql2ts/types';
// import { DEFAULT_OPTIONS } from '@gql2ts/language-typescript';

/**
 *  interface IQuery {
 *    getUser (id: string): IUser | Promise<IUser>;
 *    getString (id: string): string;
 *  }
 */

const testSchema: string = `
  type Query {
    getString(id: ID!): String!
    getUser(id: ID!): User
  }

  type User {
    id: ID!
  }

  type Schema {
    query: Query
  }
`;

interface IArgument {
  name: string;
  type: string;
}

const argsToString: (args: IArgument[]) => string = args => (
  ['{', ...args.map(({name, type}) => `${name}: ${type};`), '}'].join(' ')
);

const run: (schema: GraphQLSchema) => string = schema => {
  // const { printType }: IFromQueryOptions = DEFAULT_OPTIONS;

  const queryType: GraphQLObjectType = schema.getQueryType();
  const fields: GraphQLFieldMap<any, any> = queryType.getFields();
  Object.keys(fields).map(key => {
    const field: GraphQLField<any, any> = fields[key];
    const { args, type }: typeof field = field;
    const argsMap: any[] = args.map(({ name, type: argType }) => ({ name, type: argType }));
    console.log(`${key} (source: any, args: ${argsToString(argsMap)}, ctx: any): ${type};`);
  });

  return '';
};

export const generateResolvers: (schema: PossibleSchemaInput) => string = schema => run(schemaFromInputs(schema));

console.log(generateResolvers(testSchema));
