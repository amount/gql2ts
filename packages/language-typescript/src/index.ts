import {
  GenerateSubTypeInterface,
  WrapType,
  IFromQueryOptions,
  InputFormatter,
  QueryNamer,
  EnumFormatter,
  InterfaceAndTypeBuilder,
  InterfaceDeclarationGenerator,
  TypeJoiner,
  TypePrinter,
  NamespaceGenerator,
  InterfaceNameWithExtensions,
  EnumTypeBuilder,
  GenerateDocumentation,
  ITypeMap,
} from '@gql2ts/types';
import { buildDocumentation, filterAndJoinArray } from '@gql2ts/util';
import prettify from './typescriptPrettify';
import { pascalize } from 'humps';

export const DEFAULT_INTERFACE_DECLARATION: InterfaceDeclarationGenerator = fields => `{
${filterAndJoinArray(fields)}
}`;

export const DEFAULT_INTERFACE_BUILDER: InterfaceAndTypeBuilder = (name, body) => `interface ${name} ${body}`;
export const DEFAULT_INTERFACE_NAMER: WrapType = name => `I${pascalize(name)}`;
export const DEFAULT_TYPE_BUILDER: InterfaceAndTypeBuilder = (name, body) => `type ${name} = ${body}`;
export const DEFAULT_TYPE_JOINER: TypeJoiner = types => filterAndJoinArray(types, ' & ');
export const DEFAULT_TYPE_NAMER: WrapType = name => name;

export const interfaceExtendListToString: (extensions: string[]) => string = exts => {
  if (!exts.length) { return ''; }
  return ` extends ${filterAndJoinArray(exts, ', ')}`;
};

export const ADD_INTERFACE_EXTENSIONS: InterfaceNameWithExtensions = (opName, exts) => opName + interfaceExtendListToString(exts);
export const DEFAULT_NAME_FRAGMENT: WrapType = name => `FragmentSelectionOn${pascalize(name)}`;
export const DEFAULT_NAME_QUERY: QueryNamer = def => def.name ? pascalize(def.name.value) : 'Anonymous';

export const DEFAULT_FORMAT_INPUT: InputFormatter = (name, isOptional, type) => ADD_SEMICOLON(`${name}${isOptional ? '?:' : ':' } ${type}`);

export const DEFAULT_TYPE_MAP: ITypeMap = {
  ID: 'string',
  String: 'string',
  Boolean: 'boolean',
  Float: 'number',
  Int: 'number',
  __DEFAULT: 'any',
};

export const DEFAULT_WRAP_LIST: WrapType = type => `Array<${type}>`;
export const DEFAULT_WRAP_PARTIAL: WrapType = partial => `Partial<${partial}>`;

export const DEFAULT_TYPE_PRINTER: TypePrinter = (type, isNonNull) => isNonNull ? type : `${type} | null`;

export const DEFAULT_GENERATE_SUBTYPE_INTERFACE_NAME: GenerateSubTypeInterface = selectionName => `SelectionOn${pascalize(selectionName)}`;

export const DEFAULT_ENUM_FORMATTER: EnumFormatter = (values, documentationGenerator) => `{
  ${filterAndJoinArray(
    values.map(value => filterAndJoinArray([
      documentationGenerator(buildDocumentation(value)),
      `${value.name} = '${value.name}'`
    ])),
    ',\n'
  )}
}`;

export const DEFAULT_ENUM_TYPE_BUILDER: EnumTypeBuilder = (name, values) => `const enum ${name} ${values}`;

export const DEFAULT_ENUM_NAME_GENERATOR: WrapType = name => `${pascalize(name)}`;
export const DEFAULT_INPUT_NAME_GENERATOR: WrapType = name => `${pascalize(name)}Input`;
export const DEFAULT_EXPORT_FUNCTION: WrapType = declaration => `export ${declaration}`;
export const ADD_SEMICOLON: WrapType = str => `${str};`;

export const DEFAULT_NAMESPACE_GENERATOR: NamespaceGenerator = (namespaceName, interfaces) => `// tslint:disable
// graphql typescript definitions

declare namespace ${namespaceName} {
${interfaces}
}

// tslint:enable
`;

const fixDescriptionDocblock: (description?: string | null) => string | null | undefined = description =>
  description ? description.replace(/\n/g, '\n* ') : description;

export const DEFAULT_DOCUMENTATION_GENERATOR: GenerateDocumentation = ({ description, tags = [] }) => {
  if (!description && !tags.length) {
    return '';
  }
  const arr: Array<string | null | undefined> = [
    fixDescriptionDocblock(description),
    ...tags.map(({ tag, value }) => `@${tag} ${JSON.stringify(value)}`)
  ];
  return `
  /**
   * ${filterAndJoinArray(arr, '\n   * ')}
   */`;
};

export const DEFAULT_OPTIONS: IFromQueryOptions = {
  wrapList: DEFAULT_WRAP_LIST,
  wrapPartial: DEFAULT_WRAP_PARTIAL,
  generateSubTypeInterfaceName: DEFAULT_GENERATE_SUBTYPE_INTERFACE_NAME,
  printType: DEFAULT_TYPE_PRINTER,
  formatInput: DEFAULT_FORMAT_INPUT,
  generateFragmentName: DEFAULT_NAME_FRAGMENT,
  generateQueryName: DEFAULT_NAME_QUERY,
  formatEnum: DEFAULT_ENUM_FORMATTER,
  interfaceBuilder: DEFAULT_INTERFACE_BUILDER,
  typeBuilder: DEFAULT_TYPE_BUILDER,
  enumTypeBuilder: DEFAULT_ENUM_TYPE_BUILDER,
  typeJoiner: DEFAULT_TYPE_JOINER,
  generateInterfaceDeclaration: DEFAULT_INTERFACE_DECLARATION,
  generateEnumName: DEFAULT_ENUM_NAME_GENERATOR,
  generateTypeName: DEFAULT_TYPE_NAMER,
  generateInterfaceName: DEFAULT_INTERFACE_NAMER,
  exportFunction: DEFAULT_EXPORT_FUNCTION,
  addSemicolon: ADD_SEMICOLON,
  generateNamespace: DEFAULT_NAMESPACE_GENERATOR,
  postProcessor: prettify,
  generateInputName: DEFAULT_INPUT_NAME_GENERATOR,
  addExtensionsToInterfaceName: ADD_INTERFACE_EXTENSIONS,
  generateDocumentation: DEFAULT_DOCUMENTATION_GENERATOR,
  typeMap: DEFAULT_TYPE_MAP
};

export default DEFAULT_OPTIONS;
