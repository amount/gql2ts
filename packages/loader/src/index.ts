import * as fs from 'fs';
import gqlRun from '@gql2ts/from-query';
import { PossibleSchemaInput } from '@gql2ts/util';
import { ITypeMap, IProvidedOptions } from '@gql2ts/types';
import * as loaderUtils from 'loader-utils';
import buildDeclaration from './buildDeclaration';

interface IOptions {
  schema: PossibleSchemaInput;
  typeMap?: Partial<ITypeMap>;
  options?: Partial<IProvidedOptions>;
}

// tslint:disable:no-invalid-this
module.exports = function (source: string): void {
  if (this.cacheable) {
    this.cacheable();
  }

  const callback: (
    error: Error | null,
    content: string | Buffer,
  ) => void = this.async();

  const { schema, typeMap, options }: IOptions = loaderUtils.getOptions(this) as IOptions;

  if (!schema) {
    return callback(new Error('Schema must be provided'), source);
  }

  const declaration: string = gqlRun(schema, source, typeMap, options)
    .map(({ result }) => result)
    .join('\n');

  fs.writeFile(
    `${this.resourcePath}.d.ts`,
    buildDeclaration(declaration),
    err => callback(err, source),
  );
};
// tslint:enable:no-invalid-this
