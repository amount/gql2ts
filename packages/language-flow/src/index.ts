import { IFromQueryOptions, WrapType, EnumTypeBuilder, EnumFormatter } from '@gql2ts/types';
import { DEFAULT_OPTIONS as TS_OPTIONS } from '@gql2ts/language-typescript';
import { filterAndJoinArray } from '@gql2ts/util';

export const FLOW_WRAP_PARTIAL: WrapType = partial => `$SHAPE<${partial}>`;
export const FLOW_POST_PROCESSOR: WrapType = str => `/* @flow */

${str}
`;

export const FLOW_ENUM_TYPE: EnumTypeBuilder = (name, values) => `type ${name} = ${values}`;

export const FLOW_FORMAT_ENUM: EnumFormatter = values => filterAndJoinArray(
  values.map(({ name }) => `'${name}'`), ' | '
);

export const DEFAULT_OPTIONS: IFromQueryOptions = {
  ...TS_OPTIONS,
  wrapPartial: FLOW_WRAP_PARTIAL,
  postProcessor: FLOW_POST_PROCESSOR,
  enumTypeBuilder: FLOW_ENUM_TYPE,
  formatEnum: FLOW_FORMAT_ENUM
};
export default DEFAULT_OPTIONS;
