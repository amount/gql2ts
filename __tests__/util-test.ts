import * as fs from 'fs';
import * as utils from '../packages/util/src';
import { badWriteHandler, safeJSONParse, readFile } from '../packages/util/src/fileIO';
import schemaLanguage from './shared/simpleSchema';
import { GraphQLSchema, introspectionQuery, graphql, GraphQLNonNull, GraphQLList, GraphQLEnumType, GraphQLString, GraphQLUnionType } from 'graphql';

const builtSchema: GraphQLSchema = utils.schemaFromInputs(schemaLanguage);

describe('schema', () => {
  describe('schemaFromInputs', () => {
    it('works with schema language', () => {
      expect(builtSchema).toBeInstanceOf(GraphQLSchema);
    });

    describe('introspection query', () => {
      const promised: Promise<any> = graphql(builtSchema, introspectionQuery);

      it('works with introspected query in data key', () =>
        expect(
          promised.then(r => utils.schemaFromInputs(r))
        ).resolves.toBeInstanceOf(GraphQLSchema));

      it('works with introspected query in data key', () =>
        expect(
          promised.then(r => utils.schemaFromInputs(r.data))
        ).resolves.toBeInstanceOf(GraphQLSchema));
    });

    it('works with a schema', () => {
      expect(utils.schemaFromInputs(builtSchema)).toBeInstanceOf(GraphQLSchema);
    });

    it('throws on other', () => {
      expect(() => utils.schemaFromInputs({} as any)).toThrowError(
        'Invalid Schema Input'
      );
    });
  });
  describe('utility functions', () => {
    describe('isNonNullable', () => {
      it('returns true for a non-null', () => {
        expect(utils.isNonNullable(new GraphQLNonNull(GraphQLString))).toBe(true);
      });

      it('returns false for nullable', () => {
        expect(utils.isNonNullable(GraphQLString)).toBe(false);
      });
    });

    describe('isList', () => {
      it('returns true for a list', () => {
        expect(utils.isList(new GraphQLList(GraphQLString))).toBe(true);
      });

      it('returns false for not list', () => {
        expect(utils.isList(GraphQLString)).toBe(false);
      });
    });

    describe('isEnum', () => {
      it('returns true for an enum', () => {
        expect(utils.isEnum(new GraphQLEnumType({
          name: 'testenum',
          values: {
            test: GraphQLString
          }
        }))).toBe(true);
      });

      it('returns false for non enum', () => {
        expect(utils.isEnum(GraphQLString)).toBe(false);
      });
    });

    describe('isUnion', () => {
      it('returns true for an union', () => {
        expect(utils.isUnion(new GraphQLUnionType({
          name: 'testenum',
          types: []
        }))).toBe(true);
      });

      it('returns false for non union', () => {
        expect(utils.isUnion(GraphQLString)).toBe(false);
      });
    });

    describe('isScalar', () => {
      it('returns true for an scalar', () => {
        expect(utils.isScalar(GraphQLString)).toBe(true);
      });

      it('returns false for non scalar', () => {
        expect(utils.isScalar(new GraphQLUnionType({
          name: 'fakeenum',
          types: []
        }))).toBe(false);
      });
    });
  });
});

describe('IO stuff', () => {
  it('writes file', () => {
    const spy: jest.MockInstance<{}> = jest.spyOn(fs, 'writeFile').mockImplementation(() => null);
    utils.writeToFile('test', 'test');
    expect(spy).toHaveBeenCalledTimes(1);
    (spy as any).mockRestore();
  });

  describe('error handler', () => {
    it('does nothing if called w/ undefined', () => {
      expect(() => badWriteHandler(undefined)).not.toThrow();
    });
    it('throws if called', () => {
      expect(() => badWriteHandler(new Error('Error test'))).toThrowErrorMatchingSnapshot();
    });
  });
  it('writes file', () => {
    const spy: jest.MockInstance<{}> = jest.spyOn(fs, 'readFileSync').mockImplementation(() => '{}');
    utils.readFile('test');
    expect(spy).toHaveBeenCalledTimes(1);
    (spy as any).mockRestore();
  });

  describe('readFile', () => {
    it ('reads json', () => {
      const spy: jest.MockInstance<{}> = jest.spyOn(fs, 'readFileSync').mockImplementation(() => '{ "test": true }');
      expect(readFile('test.json')).toEqual({ test: true });
      (spy as any).mockRestore();
    });

    it ('reads strings', () => {
      const spy: jest.MockInstance<{}> = jest.spyOn(fs, 'readFileSync').mockImplementation(() => 'test');
      expect(readFile('test.gql')).toEqual('test');
      (spy as any).mockRestore();
    });
  });

  describe('safeJSONParse', () => {
    it ('falls back to string', () => {
      const input: string = 'test';
      expect(safeJSONParse(input)).toEqual(input);
    });
    it ('parses json', () => {
      const input: string = '{ "test": true }';
      expect(safeJSONParse(input)).toEqual({ test: true });
    });
  });
});

describe('parser', () => {
  describe('with description', () => {
    it ('without any tags', () => {
      expect(
        utils.buildDocumentation({
          description: 'This field is something useful'
        } as any)
      ).toMatchSnapshot();
    });

    it ('with default tag', () => {
      expect(
        utils.buildDocumentation({
          description: 'This field is something useful',
          defaultValue: 'abc'
        } as any)
      ).toMatchSnapshot();
    });

    it ('with deprecated tag w/o reason', () => {
      expect(
        utils.buildDocumentation({
          description: 'This field is something useful',
          isDeprecated: true,
          deprecationReason: undefined
        } as any)
      ).toMatchSnapshot();
    });

    it ('with deprecated tag w/ reason', () => {
      expect(
        utils.buildDocumentation({
          description: 'This field is something useful',
          isDeprecated: true,
          deprecationReason: 'Use the other field instead!'
        } as any)
      ).toMatchSnapshot();
    });
    describe('the unlikely default value & deprecated', () => {
      it ('w/o deprecation reason', () => {
        expect(
          utils.buildDocumentation({
            description: 'This field is something useful',
            defaultValue: 'abc',
            isDeprecated: true,
            deprecationReason: undefined
          } as any)
        ).toMatchSnapshot();
      });

      it ('w/ deprecation reason', () => {
        expect(
          utils.buildDocumentation({
            description: 'This field is something useful',
            defaultValue: 'abc',
            isDeprecated: true,
            deprecationReason: 'Use the other field instead!'
          } as any)
        ).toMatchSnapshot();
      });
    });
  });

  describe('without description', () => {
    it ('without any tags', () => {
      expect(
        utils.buildDocumentation({        } as any)
      ).toMatchSnapshot();
    });

    it ('with default tag', () => {
      expect(
        utils.buildDocumentation({
          defaultValue: 'abc'
        } as any)
      ).toMatchSnapshot();
    });

    it ('with deprecated tag w/o reason', () => {
      expect(
        utils.buildDocumentation({
          isDeprecated: true,
          deprecationReason: undefined
        } as any)
      ).toMatchSnapshot();
    });

    it ('with deprecated tag w/ reason', () => {
      expect(
        utils.buildDocumentation({
          isDeprecated: true,
          deprecationReason: 'Use the other field instead!'
        } as any)
      ).toMatchSnapshot();
    });
    describe('the unlikely default value & deprecated', () => {
      it ('w/o deprecation reason', () => {
        expect(
          utils.buildDocumentation({
            defaultValue: 'abc',
            isDeprecated: true,
            deprecationReason: undefined
          } as any)
        ).toMatchSnapshot();
      });

      it ('w/ deprecation reason', () => {
        expect(
          utils.buildDocumentation({
            defaultValue: 'abc',
            isDeprecated: true,
            deprecationReason: 'Use the other field instead!'
          } as any)
        ).toMatchSnapshot();
      });
    });
  });
});

describe('utils', () => {
  describe('filterAndJoinArray', () => {
    describe('with default joinChar', () => {
      it ('with empty array', () => {
        expect(utils.filterAndJoinArray([])).toMatchSnapshot();
      });
      it ('with array length 1', () => {
        expect(utils.filterAndJoinArray(['hi'])).toMatchSnapshot();
      });
      it ('with array length > 1', () => {
        expect(utils.filterAndJoinArray(['hi', 'there', 'hello'])).toMatchSnapshot();
      });
      it ('with array length > 1 w/ undefined, false, and null', () => {
        expect(utils.filterAndJoinArray(['hi', 'there', 'hello', undefined, null, false, 'woah!'])).toMatchSnapshot();
      });
      it ('with only falsy values', () => {
        expect(utils.filterAndJoinArray([undefined, null, false])).toMatchSnapshot();
      });
      it ('with falsy values & 1 truthy', () => {
        expect(utils.filterAndJoinArray([undefined, null, 'hi', false])).toMatchSnapshot();
      });
    });
    describe('with alternative joinChar', () => {
      it ('with empty array', () => {
        expect(utils.filterAndJoinArray([], '\t')).toMatchSnapshot();
      });
      it ('with array length 1', () => {
        expect(utils.filterAndJoinArray(['hi'], '\t')).toMatchSnapshot();
      });
      it ('with array length > 1', () => {
        expect(utils.filterAndJoinArray(['hi', 'there', 'hello'], '\t')).toMatchSnapshot();
      });
      it ('with array length > 1 w/ undefined, false, and null', () => {
        expect(utils.filterAndJoinArray(['hi', 'there', 'hello', undefined, null, false, 'woah!'], '\t')).toMatchSnapshot();
      });
      it ('with only falsy values', () => {
        expect(utils.filterAndJoinArray([undefined, null, false], '\t')).toMatchSnapshot();
      });
      it ('with falsy values & 1 truthy', () => {
        expect(utils.filterAndJoinArray([undefined, null, 'hi', false], '\t')).toMatchSnapshot();
      });
    });
  });
});
