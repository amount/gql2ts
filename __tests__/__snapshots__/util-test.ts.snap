// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IO stuff error handler throws if called 1`] = `"Error test"`;

exports[`parser with description the unlikely default value & deprecated w/ deprecation reason 1`] = `
Object {
  "description": "This field is something useful",
  "tags": Array [
    Object {
      "tag": "default",
      "value": "abc",
    },
    Object {
      "tag": "deprecated",
      "value": "Use the other field instead!",
    },
  ],
}
`;

exports[`parser with description the unlikely default value & deprecated w/o deprecation reason 1`] = `
Object {
  "description": "This field is something useful",
  "tags": Array [
    Object {
      "tag": "default",
      "value": "abc",
    },
    Object {
      "tag": "deprecated",
      "value": "",
    },
  ],
}
`;

exports[`parser with description with default tag 1`] = `
Object {
  "description": "This field is something useful",
  "tags": Array [
    Object {
      "tag": "default",
      "value": "abc",
    },
  ],
}
`;

exports[`parser with description with deprecated tag w/ reason 1`] = `
Object {
  "description": "This field is something useful",
  "tags": Array [
    Object {
      "tag": "deprecated",
      "value": "Use the other field instead!",
    },
  ],
}
`;

exports[`parser with description with deprecated tag w/o reason 1`] = `
Object {
  "description": "This field is something useful",
  "tags": Array [
    Object {
      "tag": "deprecated",
      "value": "",
    },
  ],
}
`;

exports[`parser with description without any tags 1`] = `
Object {
  "description": "This field is something useful",
  "tags": Array [],
}
`;

exports[`parser without description the unlikely default value & deprecated w/ deprecation reason 1`] = `
Object {
  "description": undefined,
  "tags": Array [
    Object {
      "tag": "default",
      "value": "abc",
    },
    Object {
      "tag": "deprecated",
      "value": "Use the other field instead!",
    },
  ],
}
`;

exports[`parser without description the unlikely default value & deprecated w/o deprecation reason 1`] = `
Object {
  "description": undefined,
  "tags": Array [
    Object {
      "tag": "default",
      "value": "abc",
    },
    Object {
      "tag": "deprecated",
      "value": "",
    },
  ],
}
`;

exports[`parser without description with default tag 1`] = `
Object {
  "description": undefined,
  "tags": Array [
    Object {
      "tag": "default",
      "value": "abc",
    },
  ],
}
`;

exports[`parser without description with deprecated tag w/ reason 1`] = `
Object {
  "description": undefined,
  "tags": Array [
    Object {
      "tag": "deprecated",
      "value": "Use the other field instead!",
    },
  ],
}
`;

exports[`parser without description with deprecated tag w/o reason 1`] = `
Object {
  "description": undefined,
  "tags": Array [
    Object {
      "tag": "deprecated",
      "value": "",
    },
  ],
}
`;

exports[`parser without description without any tags 1`] = `
Object {
  "description": undefined,
  "tags": Array [],
}
`;
